safenode/src/network/cmd.rs

```diff
@@ -43,11 +50,58 @@ pub enum SwarmCmd {
         resp: Response,
         channel: ResponseChannel<Response>,
     },
+    RegisterProvidedData {
+        record: Record,
+    },
+    /// Get data from the kademlia store
+    GetProvidedData {
+        key: RecordKey,
+        sender: oneshot::Sender<QueryResponse>,
+    },
 }
 
 impl SwarmDriver {
     pub(crate) fn handle_cmd(&mut self, cmd: SwarmCmd) -> Result<(), Error> {
         match cmd {
+            SwarmCmd::GetProvidedData { key, sender } => {
+                match self.swarm.behaviour_mut().kademlia.store_mut().get(&key) {
+                    Some(data) => {
+                        sender
+                            .send(crate::protocol::messages::QueryResponse::GetChunk(Ok(
+                                Chunk::new(Bytes::from(data.value.clone())),
+                            )))
+                            .map_err(|_| Error::InternalMsgChannelDropped)?;
+                    }
+                    None => {
+                        // empty response here just to test out
+                        sender
+                            .send(crate::protocol::messages::QueryResponse::GetChunk(Err(
+                                ProtocolError::ProvideRecordNotFound,
+                            )))
+                            .map_err(|_| Error::InternalMsgChannelDropped)?;
+
+                        warn!("No data found in kademlia store for key: {:?}", key);
+                    }
+                }
+            }
+            SwarmCmd::RegisterProvidedData { record } => {
+                debug!("RECORDstarttttt: {:?}", record.key);
+                let _ = self
+                    .swarm
+                    .behaviour_mut()
+                    .kademlia
+                    .start_providing(record.key.clone())?;
+
+                debug!("RECORDingg: {:?}", record.key);
+                // TODO: when do we remove records. Do we need to?
```

Comment:
When our range of responsibility has shrunk (network grows) and data we hold fall outside of it, and we are running low on space.

We would need to if we don’t want to restart due to not holding what we should.

Reply:
Indeed, this was more of a "how much of this does kad handle Q". I'll rephrase.